trace/spans/factory.go

```diff
@@ -0,0 +1,341 @@
+	// the span ID of the contained span (which will not be 0).
+	//
+	GetSpanID() uint64
+
+	// GetStart() returns the time at which the span began.  Returns a zero
+	// time if the factory is empty or the contained span was Import()ed.
+	//
+	GetStart() time.Time
+
+	// GetTracePath() returns "" if the factory is empty.  Otherwise it
+	// returns the trace's resource sub-path which will be in the form
+	// "projects/{projectID}/traces/{traceID}".
+	//
+	GetTracePath() string
+
+	// GetSpanPath() returns "" if the factory is empty.  Otherwise it
+	// returns the span's resource sub-path which will be in the form
+	// "traces/{traceID}/spans/{spanID}" where both IDs are in hexadecimal.
+	//
+	GetSpanPath() string
+
+	// GetCloudContext() returns "" if the factory is empty.  Otherwise it
+	// returns a value appropriate for the "X-Cloud-Trace-Context:" header
+	// which will be in the form "{traceID}/{spanID}" where spanID is in
+	// base 10.
+	//
+	GetCloudContext() string
+
+	// Import() returns a new factory containing a span created somewhere
+	// else.  If the traceID or spanID is invalid, then a 'nil' factory and
+	// an error are returned.  The usual reason to do this is so that you can
+	// then call NewSubSpan().
+	//
+	Import(traceID string, spanID uint64) (Factory, error)
+
+	// ImportFromHeaders() returns a new factory containing a span created
+	// somewhere else based on the "X-Cloud-Trace-Context:" header.  If the
+	// header does not contain a valid CloudContext value, then a valid but
+	// empty factory is returned.
+	//
+	ImportFromHeaders(headers http.Header) Factory
+
+	// SetHeader() sets the "X-Cloud-Trace-Context:" header if the factory
+	// is not empty.
+	//
+	SetHeader(headers http.Header)
+
+	// NewTrace() returns a new factory holding a new span, part of a new
+	// trace.  Any span held in the invoking factory is ignored.
+	//
+	NewTrace() Factory
+
+	// NewSubSpan() returns a new factory holding a new span that is a
+	// sub-span of the span contained in the invoking factory.  If the
+	// invoking factory was empty, then a failure with a stack trace is
+	// logged and a 'nil' factory is returned.
+	//
+	NewSubSpan() Factory
+
+	// NewSpan() returns a new factory holding a new span.  It does
+	// NewTrace() if the factory was empty and NewSubSpan() otherwise.
+	//
+	NewSpan() Factory
+
+	// Sets the span kind to "SERVER".  Does nothing except log a failure
+	// with a stack trace if the factory is empty.
+	//
+	SetIsServer()
+
+	// Sets the span kind to "CLIENT".  Does nothing except log a failure
+	// with a stack trace if the factory is empty.
+	//
+	SetIsClient()
+
+	// Sets the span kind to "PRODUCER".  Does nothing except log a failure
+	// with a stack trace if the factory is empty.
+	//
+	SetIsPublisher()
+
+	// Sets the span kind to "CONSUMER".  Does nothing except log a failure
+	// with a stack trace if the factory is empty.
+	//
+	SetIsSubscriber()
```

Comment:
Should this be consumer?

Reply:
I found "consumer" and "producer" a bit unclear.  The key aspect of those attributes are that they are async, ie. "pub/sub".  So calling them "publisher" and "subscriber" was a lot clearer to me.  But maybe that is more a quirk of how I think of things and not worth adding the disconnect from the terminology Google uses?

Reply:
Cool - pub/sub sounds good in that case!